### DISCUSSION TITLE
Split a large BUILD file into multiple files

### DISCUSSION THREAD

**Discussion by linuxerwang** [thought-machine/please#3328] [Q&A] [answered]
Hi,

I have a very large BUILD file for third_party/go, is it possible to split it into multiple files and let plz to load them while treat them as one file?

Thanks.

**Comment by chrisnovakovic** [comment, marked as answer]
You could probably do it with subincludes, but I wouldn't recommend it. There's a pretty clean alternative if you use `go_module`, though - you could split the module import paths on forward slash characters and have a separate intermediate directory for each component, e.g.

```
## third_party/go/github.com/agnivade/levenshtein/BUILD

go_module(
    name = "levenshtein",
    module = "github.com/agnivade/levenshtein",
    version = "v1.2.0",
)
```

You can then refer to the module as `//third_party/go/github.com/agnivade/levenshtein`. You'd have a lot of `BUILD` files this way, but each one would be tiny.

It doesn't work out as cleanly as this if you use `go_repo`, because `go_repo` defines a subrepo inside the package containing the `go_repo` target definition whose name is derived from the `module` parameter, and the module is built inside that subrepo. In this case:

```
## third_party/go/github.com/agnivade/levenshtein/BUILD

go_repo(
    name = "levenshtein",
    module = "github.com/agnivade/levenshtein",
    version = "v1.2.0",
)
```

you'd end up with the module at `///third_party/go/github.com/agnivade/levenshtein/github.com_agnivade_levenshtein//:levenshtein` - rather unwieldy.

There shouldn't really be a performance problem associated with `third_party/go/BUILD` being so large, though - if the problem is one of maintenance, you might want to give [Puku](https://github.com/please-build/puku) a try.
